### PR TITLE
FINERACT-2102: Loan Product enableDownPayment overrided in the Loan a…

### PR DESCRIPTION
**Pull request by alberto-art3ch**
…pplication

## Description

There is need to decide dynamically the loan is downpayment loan or not under the same product constraints. The change here to allow override previously configured `enableDownPayment` field

[FINERACT-2102](https://issues.apache.org/jira/browse/FINERACT-2102)

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanAssembler.java [hunk 1]
```diff
@@ -835,6 +835,12 @@ public Map<String, Object> updateFrom(JsonCommand command, Loan loan) {
             loan.updateLoanRates(rateAssembler.fromParsedJson(command.parsedJson()));
         }
 
+        if (command.hasParameter(LoanProductConstants.ENABLE_DOWN_PAYMENT)) {
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanAssembler.java]: Please follow the same logic as it was done when loan was submitted:
- You can disable down payment on loan, when the it was allowed on the loan product, but not the otherwise! It should fail with validation when not allowed action was tried:
- user tries to enable down payment but it was not configured on loan product!
Ignoring the field silently that was sent is wrong!

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/AdvancedPaymentAllocationLoanRepaymentScheduleTest.java [hunk 2]
```diff
@@ -4113,6 +4113,45 @@ public void uc137() {
         });
     }
 
+    // UC138: Advanced payment allocation, with DownPayment enabled in Loan Product but disabled in Loan application
```
Comment (adamsaghy) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/AdvancedPaymentAllocationLoanRepaymentScheduleTest.java:4118]: Please add a test case for update loan application!
Please add negative test cases to check enabling down payment is not allowed if it was not allowed on the loan product!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/service/LoanScheduleAssembler.java [hunk 3]
```diff
@@ -466,7 +466,11 @@ private LoanApplicationTerms assembleLoanApplicationTermsFrom(final JsonElement
         final boolean isPrincipalCompoundingDisabledForOverdueLoans = this.configurationDomainService
                 .isPrincipalCompoundingDisabledForOverdueLoans();
 
-        final boolean isDownPaymentEnabled = loanProduct.getLoanProductRelatedDetail().isEnableDownPayment();
+        boolean isDownPaymentEnabled = loanProduct.getLoanProductRelatedDetail().isEnableDownPayment();
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/service/LoanScheduleAssembler.java:469]: Please add validation error when user tries to enable down payment but it was not configured on loan product!
Ignoring the field silently that was sent is wrong!

**Review by adamsaghy — CHANGES_REQUESTED** [review]
Please kindly check my review!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/service/LoanScheduleAssembler.java [hunk 4]
```diff
@@ -466,7 +467,9 @@ private LoanApplicationTerms assembleLoanApplicationTermsFrom(final JsonElement
         final boolean isPrincipalCompoundingDisabledForOverdueLoans = this.configurationDomainService
                 .isPrincipalCompoundingDisabledForOverdueLoans();
 
-        final boolean isDownPaymentEnabled = loanProduct.getLoanProductRelatedDetail().isEnableDownPayment();
+        final boolean isDownPaymentEnabled = validateDownPaymentAttribute(loanProduct.getLoanProductRelatedDetail().isEnableDownPayment(),
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/service/LoanScheduleAssembler.java]: Please move this into the loan validator!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/service/LoanScheduleAssembler.java [hunk 5]
```diff
@@ -1379,6 +1382,21 @@ public void updateLoanApplicationAttributes(JsonCommand command, Loan loan, Map<
             changes.put(LoanProductConstants.IS_EQUAL_AMORTIZATION_PARAM, newValue);
             loanProductRelatedDetail.setEqualAmortization(newValue);
         }
+
+        final boolean isDownPaymentEnabled = validateDownPaymentAttribute(
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/service/LoanScheduleAssembler.java]: Please move this into the loan validator!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/service/LoanScheduleAssembler.java [hunk 6]
```diff
@@ -1379,6 +1382,21 @@ public void updateLoanApplicationAttributes(JsonCommand command, Loan loan, Map<
...
+                loan.loanProduct().getLoanProductRelatedDetail().isEnableDownPayment(), command.parsedJson());
+        if (isDownPaymentEnabled != loanProductRelatedDetail.isEnableDownPayment()) {
+            changes.put(LoanProductConstants.ENABLE_DOWN_PAYMENT, isDownPaymentEnabled);
+            loanProductRelatedDetail.setEnableDownPayment(isDownPaymentEnabled);
+            loanProductRelatedDetail.setEnableAutoRepaymentForDownPayment(isDownPaymentEnabled);
+            // Reset the other DownPayment attributes
+            if (!isDownPaymentEnabled) {
+                loanProductRelatedDetail.setDisbursedAmountPercentageForDownPayment(null);
+            } else {
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/service/LoanScheduleAssembler.java:1395]: This case cannot happen due to the validation restrictions.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/service/LoanScheduleAssembler.java [hunk 7]
```diff
@@ -1379,6 +1382,21 @@ public void updateLoanApplicationAttributes(JsonCommand command, Loan loan, Map<
             changes.put(LoanProductConstants.IS_EQUAL_AMORTIZATION_PARAM, newValue);
             loanProductRelatedDetail.setEqualAmortization(newValue);
         }
+
+        final boolean isDownPaymentEnabled = validateDownPaymentAttribute(
+                loan.loanProduct().getLoanProductRelatedDetail().isEnableDownPayment(), command.parsedJson());
+        if (isDownPaymentEnabled != loanProductRelatedDetail.isEnableDownPayment()) {
+            changes.put(LoanProductConstants.ENABLE_DOWN_PAYMENT, isDownPaymentEnabled);
+            loanProductRelatedDetail.setEnableDownPayment(isDownPaymentEnabled);
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/service/LoanScheduleAssembler.java]: Its more straightforward to set the values as "false" instead of assigning the value of the `isDownPaymentEnabled` which cannot be anything else but false.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/service/LoanScheduleAssembler.java [hunk 8]
```diff
@@ -1446,4 +1464,21 @@ public Pair<Loan, Map<String, Object>> assembleLoanApproval(AppUser currentUser,
 
         return Pair.of(loan, actualChanges);
     }
+
+    private boolean validateDownPaymentAttribute(final boolean isDownPaymentEnabledInLoanProduct, final JsonElement element) {
+        boolean isDownPaymentEnabled = isDownPaymentEnabledInLoanProduct;
+        if (!isDownPaymentEnabledInLoanProduct) {
+            if (this.fromApiJsonHelper.parameterExists(LoanProductConstants.ENABLE_DOWN_PAYMENT, element)) {
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/service/LoanScheduleAssembler.java]: This is wrong. We should only throw exception if the the down payment is disabled on the loan product (which is correctly checked) and when user tries to set down payment to be enabled during loan submission (this is the wrongly checked part)! If user provide enableDownPayment=false then nothing should happen: No violation!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/service/LoanScheduleAssembler.java [hunk 9]
```diff
@@ -1446,4 +1464,21 @@ public Pair<Loan, Map<String, Object>> assembleLoanApproval(AppUser currentUser,
 
         return Pair.of(loan, actualChanges);
     }
+
+    private boolean validateDownPaymentAttribute(final boolean isDownPaymentEnabledInLoanProduct, final JsonElement element) {
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/service/LoanScheduleAssembler.java]: Please move the validations into the Validator class!

**Review by adamsaghy — CHANGES_REQUESTED** [review]
Please kindly see my review!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/service/LoanScheduleAssembler.java (hunk 8 above)
Comment (alberto-art3ch) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/service/LoanScheduleAssembler.java]: Done! Logic updated

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/service/LoanScheduleAssembler.java (hunk 7 above)
Comment (alberto-art3ch) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/service/LoanScheduleAssembler.java]: We need to consider a couple of scenarios:
1. LP with DownPayment enabled and originally the LA with DownPayment enabled, then LA disable DownPayment, and
2.LP with DownPayment enabled and originally the LA with DownPayment disabled, then LA enable DownPayment

This is why we are validating If we have a change in the new value compared with the old value

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/service/LoanScheduleAssembler.java (hunk 5 above)
Comment (alberto-art3ch) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/service/LoanScheduleAssembler.java]: Done!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/service/LoanScheduleAssembler.java (hunk 3 above)
Comment (alberto-art3ch) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/service/LoanScheduleAssembler.java:469]: Done! new validation method implemented in the LoanScheduleValidator

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/AdvancedPaymentAllocationLoanRepaymentScheduleTest.java (hunk 2 above)
Comment (alberto-art3ch) [reply integration-tests/src/test/java/org/apache/fineract/integrationtests/AdvancedPaymentAllocationLoanRepaymentScheduleTest.java:4118]: Done! ITC added

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanAssembler.java (hunk 1 above)
Comment (alberto-art3ch) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanAssembler.java]: Done! new validation method implemented in the LoanScheduleValidator

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/data/LoanSummaryData.java [hunk 10]
```diff
@@ -91,9 +91,14 @@ public class LoanSummaryData {
     private BigDecimal totalCreditBalanceRefundReversed;
     private BigDecimal totalRepaymentTransaction;
     private BigDecimal totalRepaymentTransactionReversed;
+    private BigDecimal totalInterestRefund;
```
Comment (adamsaghy) [inline fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/data/LoanSummaryData.java]: totalInterestRefund is not yet calculated. Please remove till it got implemented fully.

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/data/LoanSummaryData.java [hunk 11]
```diff
@@ -91,9 +91,14 @@ public class LoanSummaryData {
...
+    private BigDecimal totalInterestPaymentWaiver;
     private final Long chargeOffReasonId;
     private final String chargeOffReason;
 
+    private BigDecimal totalUnpaidAccruedDueInterest;
```
Comment (adamsaghy) [inline fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/data/LoanSummaryData.java]: totalUnpaidAccruedDueInterest and totalUnpaidAccruedNotDueInterest are not set with value. Please compute their value or remove them till it got implemented fully.

File: fineract-progressive-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/transactionprocessor/impl/ChargeOrTransaction.java [hunk 12]
```diff
@@ -58,7 +58,7 @@ private LocalDate getEffectiveDate() {
     }
 
     private boolean isBackdatedCharge() {
-        return loanCharge.get().getDueDate().isBefore(loanCharge.get().getSubmittedOnDate());
+        return (loanCharge.get().getDueDate() != null && loanCharge.get().getDueDate().isBefore(loanCharge.get().getSubmittedOnDate()));
```
Comment (adamsaghy) [inline fineract-progressive-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/transactionprocessor/impl/ChargeOrTransaction.java]: This change is unrelated. If this is a fix for something, please raise it in a new PR!

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/LoanTransactionType.java [hunk 13]
```diff
@@ -63,8 +63,7 @@ public enum LoanTransactionType {
     DOWN_PAYMENT(28, "loanTransactionType.downPayment"), //
     REAGE(29, "loanTransactionType.reAge"), //
     REAMORTIZE(30, "loanTransactionType.reAmortize"), //
-    INTEREST_PAYMENT_WAIVER(31, "loanTransactionType.interestPaymentWaiver"), //
-    ;
+    INTEREST_PAYMENT_WAIVER(31, "loanTransactionType.interestPaymentWaiver"); //
```
Comment (adamsaghy) [inline fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/LoanTransactionType.java]: This is unrelated change. Please remove it

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoansApiResourceSwagger.java [hunk 14]
```diff
@@ -636,6 +636,10 @@ private GetLoansLoanIdFeeFrequency() {}
             @Schema(example = "0.000000")
             public Double totalRepaymentTransaction;
             @Schema(example = "0.000000")
+            public Double totalInterestRefund;
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoansApiResourceSwagger.java]: totalInterestRefund is not set properly. Please remove till!

**Review by adamsaghy — CHANGES_REQUESTED** [review]
Seems like you mixed together different PRs! Please remove the unrelated changes!

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/delinquency/service/LoanDelinquencyDomainServiceImpl.java [hunk 15]
```diff
@@ -64,13 +64,22 @@ public CollectionData getOverdueCollectionData(final Loan loan, List<LoanDelinqu
             return CollectionData.template();
         }
 
+        BigDecimal delinquentPrincipal = BigDecimal.ZERO;
```
Comment (adamsaghy) [inline fineract-loan/src/main/java/org/apache/fineract/portfolio/delinquency/service/LoanDelinquencyDomainServiceImpl.java]: unrelated changes!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/service/LoanScheduleAssembler.java [hunk 16]
```diff
@@ -466,7 +468,9 @@ private LoanApplicationTerms assembleLoanApplicationTermsFrom(final JsonElement
         final boolean isPrincipalCompoundingDisabledForOverdueLoans = this.configurationDomainService
                 .isPrincipalCompoundingDisabledForOverdueLoans();
 
-        final boolean isDownPaymentEnabled = loanProduct.getLoanProductRelatedDetail().isEnableDownPayment();
+        final boolean isDownPaymentEnabled = loanScheduleValidator
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/service/LoanScheduleAssembler.java]: Validations are called when loan got submitted or updated:
`LoanApplicationWritePlatformServiceJpaRepositoryImpl.submitApplication`
and
`LoanApplicationWritePlatformServiceJpaRepositoryImpl.modifyApplication`

Please move the validations into the `loanApplicationValidator.validateForCreate` and `loanApplicationValidator.validateForModify`

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/service/LoanScheduleAssembler.java (hunk 7 above)
Comment (adamsaghy) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/service/LoanScheduleAssembler.java:1392]: But then you need to check whether the provided `enableDownPayment` is different than the one is already set on the loan, not on the loan product. 

The following situation will have no effect and the down payment remains disabled at the end.
LP: enable down payment is TRUE
Submit loan: enable down payment is FALSE
Update loan: enabled down payment is TRUE -> Down payment remains disabled hence we are comparing the newly provided value with the LP value.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanScheduleValidator.java [hunk 17]
```diff
@@ -176,4 +178,24 @@ private void validateRepaymentsStartingFromDateIsAfterDisbursementDate(final Lis
             dataValidationErrors.add(error);
         }
     }
+
+    public boolean validateDownPaymentAttribute(final boolean isDownPaymentEnabledInLoanProduct, final JsonElement element) {
+        boolean isDownPaymentEnabled = isDownPaymentEnabledInLoanProduct;
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanScheduleValidator.java]: This method is overcomplicated:
```
public void validateDownPaymentAttribute(final boolean isDownPaymentEnabledInLoanProduct, final JsonElement element) {
      Boolean inputIsDownPaymentEnabled = this.fromApiJsonHelper.extractBooleanNamed(LoanProductConstants.ENABLE_DOWN_PAYMENT, element);
      if(!isDownPaymentEnabledInLoanProduct && Boolean.TRUE.equals(inputIsDownPaymentEnabled)) {
               throw new GeneralPlatformDomainRuleException("error.msg.downpayment.is.not.enabled.in.loan.product",
                        "The Loan can not override the downpayment flag because in the Loan Product the downpayment is disabled",
                        inputIsDownPaymentEnabled);
                        }
}
```

What do you think?

P.S.: You can consider passing the whole loan product as parameter and the validation method will decide what it is using from. Similarly the whole command passed as 2nd parameter

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/service/LoanScheduleAssembler.java [hunk 18]
```diff
@@ -1379,6 +1383,22 @@ public void updateLoanApplicationAttributes(JsonCommand command, Loan loan, Map<
             changes.put(LoanProductConstants.IS_EQUAL_AMORTIZATION_PARAM, newValue);
             loanProductRelatedDetail.setEqualAmortization(newValue);
         }
+
+        final boolean isDownPaymentEnabled = loanScheduleValidator
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/service/LoanScheduleAssembler.java]: We only need the incoming `isDownPaymentEnabled` and if it is different than the one on the loan actually set then either set the LP config or reset it (based on we enable or disable it).

File: fineract-progressive-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/transactionprocessor/impl/ChargeOrTransaction.java (hunk 12 above)
Comment (alberto-art3ch) [reply fineract-progressive-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/transactionprocessor/impl/ChargeOrTransaction.java]: Done! removed, It belongs to an issue when you have a Loan Installment Fee

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/data/LoanSummaryData.java (hunk 11 above)
Comment (alberto-art3ch) [reply fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/data/LoanSummaryData.java]: Done! Removed

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/data/LoanSummaryData.java (hunk 10 above)
Comment (alberto-art3ch) [reply fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/data/LoanSummaryData.java]: Done! Removed

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/service/LoanScheduleAssembler.java (hunk 9 above)
Comment (alberto-art3ch) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/service/LoanScheduleAssembler.java]: Done! moved to the validator

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/LoanTransactionType.java (hunk 13 above)
Comment (alberto-art3ch) [reply fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/LoanTransactionType.java]: Done! Removed

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoansApiResourceSwagger.java (hunk 14 above)
Comment (alberto-art3ch) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoansApiResourceSwagger.java]: Done! Removed

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/delinquency/service/LoanDelinquencyDomainServiceImpl.java [hunk 19]
```diff
@@ -110,6 +110,7 @@ public CollectionData getOverdueCollectionData(final Loan loan, List<LoanDelinqu
             collectionData.setDelinquentDate(overdueSinceDate);
         }
         collectionData.setDelinquentAmount(outstandingAmount);
+
```
Comment (adamsaghy) [inline fineract-loan/src/main/java/org/apache/fineract/portfolio/delinquency/service/LoanDelinquencyDomainServiceImpl.java]: Unrelated changes, please remove!

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/data/LoanSummaryData.java [hunk 20]
```diff
@@ -110,6 +111,8 @@ public static LoanSummaryData withTransactionAmountsSummary(final LoanSummaryDat
         BigDecimal totalCreditBalanceRefundReversed = BigDecimal.ZERO;
         BigDecimal totalRepaymentTransaction = BigDecimal.ZERO;
         BigDecimal totalRepaymentTransactionReversed = BigDecimal.ZERO;
+        BigDecimal totalInterestRefund = BigDecimal.ZERO;
```
Comment (adamsaghy) [inline fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/data/LoanSummaryData.java]: Unrelated changes, please remove!

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/data/LoanSummaryData.java [hunk 21]
```diff
@@ -131,6 +134,8 @@ public static LoanSummaryData withTransactionAmountsSummary(final LoanSummaryDat
                     loanTransactions);
             totalRepaymentTransaction = computeTotalRepaymentTransactionAmount(loanTransactions);
             totalRepaymentTransactionReversed = computeTotalAmountForReversedTransactions(LoanTransactionType.REPAYMENT, loanTransactions);
+            totalInterestPaymentWaiver = computeTotalAmountForNonReversedTransactions(LoanTransactionType.INTEREST_PAYMENT_WAIVER,
```
Comment (adamsaghy) [inline fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/data/LoanSummaryData.java]: Unrelated changes, please remove!

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/data/LoanSummaryData.java [hunk 22]
```diff
@@ -163,7 +168,7 @@ public static LoanSummaryData withTransactionAmountsSummary(final LoanSummaryDat
                 .totalChargeAdjustment(totalChargeAdjustment).totalChargeAdjustmentReversed(totalChargeAdjustmentReversed)
                 .totalChargeback(totalChargeback).totalCreditBalanceRefund(totalCreditBalanceRefund)
                 .totalCreditBalanceRefundReversed(totalCreditBalanceRefundReversed).totalRepaymentTransaction(totalRepaymentTransaction)
-                .totalRepaymentTransactionReversed(totalRepaymentTransactionReversed).build();
```
Comment (adamsaghy) [inline fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/data/LoanSummaryData.java]: Unrelated changes, please remove!

File: fineract-progressive-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/transactionprocessor/impl/ChargeOrTransaction.java [hunk 23]
```diff
@@ -58,7 +58,7 @@ private LocalDate getEffectiveDate() {
     }
 
     private boolean isBackdatedCharge() {
-        return loanCharge.get().getDueDate().isBefore(loanCharge.get().getSubmittedOnDate());
+        return (loanCharge.get().getDueDate().isBefore(loanCharge.get().getSubmittedOnDate()));
```
Comment (adamsaghy) [inline fineract-progressive-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/transactionprocessor/impl/ChargeOrTransaction.java]: Unrelated changes, please remove!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoansApiResourceSwagger.java [hunk 24]
```diff
@@ -1004,6 +1004,14 @@ private GetLoansLoanIdDelinquencySummary() {}
             public LocalDate delinquentDate;
             @Schema(example = "100.000000")
             public Double delinquentAmount;
+            @Schema(example = "80.000000")
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoansApiResourceSwagger.java]: Unrelated changes, please remove!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoansApiResourceSwagger.java [hunk 25]
```diff
@@ -1245,6 +1253,8 @@ private PostLoansRequest() {}
         public String loanScheduleProcessingType;
         @Schema(example = "false")
         public Boolean enableInstallmentLevelDelinquency;
+        @Schema(example = "false")
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoansApiResourceSwagger.java:1256]: Unrelated changes, please remove!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoansApiResourceSwagger.java [hunk 26]
```diff
@@ -1394,6 +1404,8 @@ private PutLoansLoanIdRequest() {}
...
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoansApiResourceSwagger.java:1407]: Unrelated changes, please remove!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/service/LoanScheduleAssembler.java (hunk 7 above)
Comment (adamsaghy) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/service/LoanScheduleAssembler.java]: You are checking the loan product configuration, but what you should check is only the configuration on loan. This method is covering the update loan application part, we already have something configured on the loan, thats the only thing that should be checked! 

If on the loan product it was enabled down payment, but it was disabled when the loan was submitted and now when the loan is got updated, we want to enable again, you should not check what was set on the loan product, but check whether on the loan level what was set and now during the update if we wanna change it and enable again, then we can set from the loan product the down payment configurations on the loan level again.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/service/LoanScheduleAssembler.java [hunk 27]
```diff
@@ -1379,6 +1383,22 @@ public void updateLoanApplicationAttributes(JsonCommand command, Loan loan, Map<
...
+                .validateDownPaymentAttribute(loan.loanProduct().getLoanProductRelatedDetail().isEnableDownPayment(), command.parsedJson());
+        // If the value is different, new value compared with the previous value
+        if (isDownPaymentEnabled != loanProductRelatedDetail.isEnableDownPayment()) {
+            changes.put(LoanProductConstants.ENABLE_DOWN_PAYMENT, isDownPaymentEnabled);
+            loanProductRelatedDetail.setEnableDownPayment(isDownPaymentEnabled);
+            loanProductRelatedDetail.setEnableAutoRepaymentForDownPayment(isDownPaymentEnabled);
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/service/LoanScheduleAssembler.java]: Just because down payment was enabled, it does not mean auto repayment for down payment should be enabled as well. It should be set based on what was configured on the loan product

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/service/LoanScheduleAssembler.java (hunk 16 above)
Comment (alberto-art3ch) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/service/LoanScheduleAssembler.java]: Done! All the validations are in the `LoanScheduleValidator` class

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/service/LoanScheduleAssembler.java (hunk 18 above)
Comment (alberto-art3ch) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/service/LoanScheduleAssembler.java]: Done! In the `LoanScheduleAssembler` is only the input read If exists, else we set the LP value

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/service/LoanScheduleAssembler.java (hunk 27 above)
Comment (alberto-art3ch) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/service/LoanScheduleAssembler.java]: Done! We are just changing the `enableDownPayment` attribute

**Review by adamsaghy — APPROVED** [review]
LGTM
